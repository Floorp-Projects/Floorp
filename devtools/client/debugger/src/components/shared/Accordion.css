/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.accordion {
  background-color: var(--theme-sidebar-background);
  width: 100%;
  list-style-type: none;
  padding: 0px;
  margin-top: 0px;
}

.accordion ._header {
  background-color: var(--theme-accordion-header-background);
  border-bottom: 1px solid var(--theme-splitter-color);
  display: flex;
  font-size: 12px;
  line-height: calc(16 / 12);
  padding: 4px 6px;
  width: 100%;
  align-items: center;
  margin: 0px;
  font-weight: normal;
  cursor: default;
  user-select: none;
}

.accordion ._header:hover {
  background-color: var(--theme-accordion-header-hover);
}

.accordion ._header .arrow {
  margin-inline-end: 4px;
}

.accordion ._header .header-label {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: var(--theme-toolbar-color);
}

.accordion ._header .header-buttons {
  display: flex;
  margin-inline-start: auto;
}

.accordion ._header .header-buttons button {
  color: var(--theme-body-color);
  border: none;
  background: none;
  padding: 0;
  margin: 0 2px;
  width: 16px;
  height: 16px;
}

.accordion ._header .header-buttons button::-moz-focus-inner {
  border: none;
}

.accordion ._header .header-buttons button .img {
  display: block;
}

.accordion ._content {
  border-bottom: 1px solid var(--theme-splitter-color);
  font-size: var(--theme-body-font-size);
}

.accordion div:last-child ._content {
  border-bottom: none;
}
