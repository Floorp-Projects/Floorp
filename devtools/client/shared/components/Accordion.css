/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* Accordion */

.accordion {
  width: 100%;
  padding: 0;
  margin: 0;
  list-style-type: none;
  /* Accordion root has tabindex="-1" to get focus programatically.
   * This can give it a focus outline when clicked, which we don't want.
   * The container itself is not in the focus order at all. */
  outline: none;
  background-color: var(--theme-sidebar-background);
}

.accordion-header {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  /* Reserve 1px for the border */
  min-height: calc(var(--theme-toolbar-height) + 1px);
  margin: 0;
  border-bottom: 1px solid var(--theme-splitter-color);
  padding: 2px 4px;
  font-size: inherit;
  font-weight: normal;
  user-select: none;
  cursor: default;
  background-color: var(--theme-accordion-header-background);
}

.accordion-header:hover {
  background-color: var(--theme-accordion-header-hover);
}

/*
  Arrow should be a bit closer to the text than to the start edge:
  - total distance between text and start edge = 20px
  - arrow width = 10px
  - distance between arrow and start edge = 6px
  - distance between arrow and text = 4px
*/
.accordion-header .theme-twisty {
  display: inline-block;
  flex: none;
  width: 10px;
  height: 10px;
  margin-inline-start: 2px;
  margin-inline-end: 4px;
  pointer-events: none;
}

.accordion-header-label {
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 12px;
  line-height: 16px;
  color: var(--theme-toolbar-color);
}

.accordion-header-buttons {
  flex: none;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  max-width: 50%;
  margin-inline-start: auto;
  padding-inline-start: 4px;
}

.accordion-content {
  overflow: auto;
  border-bottom: 1px solid var(--theme-splitter-color);
}

.accordion-content[hidden] {
  display: none;
}

.accordion-item:last-child > .accordion-content {
  border-bottom: none;
}
