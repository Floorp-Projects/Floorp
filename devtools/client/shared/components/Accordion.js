/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

"use strict";

const {
  Component,
  createElement,
} = require("devtools/client/shared/vendor/react");
const PropTypes = require("devtools/client/shared/vendor/react-prop-types");
const {
  ul,
  li,
  h2,
  div,
  span,
} = require("devtools/client/shared/vendor/react-dom-factories");

class Accordion extends Component {
  static get propTypes() {
    return {
      className: PropTypes.string,
      // A list of all items to be rendered using an Accordion component.
      items: PropTypes.arrayOf(
        PropTypes.shape({
          buttons: PropTypes.arrayOf(PropTypes.object),
          className: PropTypes.string,
          component: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
          componentProps: PropTypes.object,
          contentClassName: PropTypes.string,
          header: PropTypes.string.isRequired,
          id: PropTypes.string.isRequired,
          onToggle: PropTypes.func,
          // Determines the initial open state of the accordion item
          opened: PropTypes.bool.isRequired,
          // Enables dynamically changing the open state of the accordion
          // on update.
          shouldOpen: PropTypes.func,
        })
      ).isRequired,
    };
  }

  constructor(props) {
    super(props);

    this.state = {
      opened: {},
    };

    this.onHeaderClick = this.onHeaderClick.bind(this);
    this.onHeaderKeyDown = this.onHeaderKeyDown.bind(this);
    this.setInitialState = this.setInitialState.bind(this);
    this.updateCurrentState = this.updateCurrentState.bind(this);
  }

  componentDidMount() {
    this.setInitialState();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.items !== this.props.items) {
      this.updateCurrentState();
    }
  }

  setInitialState() {
    /**
     * Add initial data to the `state.opened` map.
     * This happens only on initial mount of the accordion.
     */
    const newItems = this.props.items.filter(
      ({ id }) => typeof this.state.opened[id] !== "boolean"
    );

    if (newItems.length) {
      const everOpened = { ...this.state.everOpened };
      const opened = { ...this.state.opened };
      for (const item of newItems) {
        everOpened[item.id] = item.opened;
        opened[item.id] = item.opened;
      }

      this.setState({ everOpened, opened });
    }
  }

  updateCurrentState() {
    /**
     * Updates the `state.opened` map based on the
     * new items that have been added and those that
     * `item.shouldOpen()` has changed. This happens
     * on each update.
     */
    const updatedItems = this.props.items.filter(item => {
      const notExist = typeof this.state.opened[item.id] !== "boolean";
      if (typeof item.shouldOpen == "function") {
        const currentState = this.state.opened[item.id];
        return notExist || currentState !== item.shouldOpen(item, currentState);
      }
      return notExist;
    });

    if (updatedItems.length) {
      const everOpened = { ...this.state.everOpened };
      const opened = { ...this.state.opened };
      for (const item of updatedItems) {
        let itemOpen = item.opened;
        if (typeof item.shouldOpen == "function") {
          itemOpen = item.shouldOpen(item, itemOpen);
        }
        everOpened[item.id] = itemOpen;
        opened[item.id] = itemOpen;
      }
      this.setState({ everOpened, opened });
    }
  }

  /**
   * @param {Event} event Click event.
   * @param {Object} item The item to be collapsed/expanded.
   */
  onHeaderClick(event, item) {
    event.preventDefault();
    // In the Browser Toolbox's Inspector/Layout view, handleHeaderClick is
    // called twice unless we call stopPropagation, making the accordion item
    // open-and-close or close-and-open
    event.stopPropagation();
    this.toggleItem(item);
  }

  /**
   * @param {Event} event Keyboard event.
   * @param {Object} item The item to be collapsed/expanded.
   */
  onHeaderKeyDown(event, item) {
    if (event.key === " " || event.key === "Enter") {
      event.preventDefault();
      this.toggleItem(item);
    }
  }

  /**
   * Expand or collapse an accordion list item.
   * @param  {Object} item The item to be collapsed or expanded.
   */
  toggleItem(item) {
    const opened = !this.state.opened[item.id];

    this.setState({
      everOpened: {
        ...this.state.everOpened,
        [item.id]: true,
      },
      opened: {
        ...this.state.opened,
        [item.id]: opened,
      },
    });

    if (typeof item.onToggle === "function") {
      item.onToggle(opened, item);
    }
  }

  renderItem(item) {
    const {
      buttons,
      className = "",
      component,
      componentProps = {},
      contentClassName = "",
      header,
      id,
    } = item;

    const headerId = `${id}-header`;
    const opened = this.state.opened[id];
    let itemContent;

    // Only render content if the accordion item is open or has been opened once before.
    // This saves us rendering complex components when users are keeping
    // them closed (e.g. in Inspector/Layout) or may not open them at all.
    if (this.state.everOpened && this.state.everOpened[id]) {
      if (typeof component === "function") {
        itemContent = createElement(component, componentProps);
      } else if (typeof component === "object") {
        itemContent = component;
      }
    }

    return li(
      {
        key: id,
        id,
        className: `accordion-item ${
          opened ? "accordion-open" : ""
        } ${className} `.trim(),
        "aria-labelledby": headerId,
      },
      h2(
        {
          id: headerId,
          className: "accordion-header",
          tabIndex: 0,
          "aria-expanded": opened,
          // If the header contains buttons, make sure the heading name only
          // contains the "header" text and not the button text
          "aria-label": header,
          onKeyDown: event => this.onHeaderKeyDown(event, item),
          onClick: event => this.onHeaderClick(event, item),
        },
        span({
          className: `theme-twisty${opened ? " open" : ""}`,
          role: "presentation",
        }),
        span(
          {
            className: "accordion-header-label",
          },
          header
        ),
        buttons &&
          span(
            {
              className: "accordion-header-buttons",
              role: "presentation",
            },
            buttons
          )
      ),
      div(
        {
          className: `accordion-content ${contentClassName}`.trim(),
          hidden: !opened,
          role: "presentation",
        },
        itemContent
      )
    );
  }

  render() {
    return ul(
      {
        className:
          "accordion" +
          (this.props.className ? ` ${this.props.className}` : ""),
        tabIndex: -1,
      },
      this.props.items.map(item => this.renderItem(item))
    );
  }
}

module.exports = Accordion;
