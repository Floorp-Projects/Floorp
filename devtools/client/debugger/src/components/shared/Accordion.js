/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { cloneElement, Component } from "react";
import AccessibleImage from "./AccessibleImage";

import "./Accordion.css";

class Accordion extends Component {
  handleHeaderClick(i) {
    const item = this.props.items[i];
    const opened = !item.opened;
    item.opened = opened;

    if (item.onToggle) {
      item.onToggle(opened);
    }

    // We force an update because otherwise the accordion
    // would not re-render
    this.forceUpdate();
  }

  onHandleHeaderKeyDown(e, i) {
    if (e && (e.key === " " || e.key === "Enter")) {
      this.handleHeaderClick(i);
    }
  }

  renderContainer = (item, i) => {
    const { opened } = item;

    return (
      <li className={item.className} key={i}>
        <h2
          className="_header"
          tabIndex="0"
          onKeyDown={e => this.onHandleHeaderKeyDown(e, i)}
          onClick={() => this.handleHeaderClick(i)}
        >
          <AccessibleImage className={`arrow ${opened ? "expanded" : ""}`} />
          <span className="header-label">{item.header}</span>
          {item.buttons ? (
            <div className="header-buttons" tabIndex="-1">
              {item.buttons}
            </div>
          ) : null}
        </h2>
        {opened && (
          <div className="_content">
            {cloneElement(item.component, item.componentProps || {})}
          </div>
        )}
      </li>
    );
  };
  render() {
    return (
      <ul className="accordion">
        {this.props.items.map(this.renderContainer)}
      </ul>
    );
  }
}

export default Accordion;
