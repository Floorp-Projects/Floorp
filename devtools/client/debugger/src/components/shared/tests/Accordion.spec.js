/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";

import Accordion from "../Accordion";

describe("Accordion", () => {
  const testItems = [
    {
      header: "Test Accordion Item 1",
      className: "accordion-item-1",
      component: <div />,
      opened: false,
      onToggle: jest.fn(),
    },
    {
      header: "Test Accordion Item 2",
      className: "accordion-item-2",
      component: <div />,
      buttons: <button />,
      opened: false,
      onToggle: jest.fn(),
    },
    {
      header: "Test Accordion Item 3",
      className: "accordion-item-3",
      component: <div />,
      opened: true,
      onToggle: jest.fn(),
    },
  ];
  const wrapper = shallow(<Accordion items={testItems} />);
  it("basic render", () => expect(wrapper).toMatchSnapshot());
  wrapper.find(".accordion-item-1 ._header").simulate("click");
  it("handleClick and onToggle", () =>
    expect(testItems[0].onToggle).toHaveBeenCalledWith(true));
});
