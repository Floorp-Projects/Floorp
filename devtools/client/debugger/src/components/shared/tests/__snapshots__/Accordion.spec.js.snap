// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion basic render 1`] = `
<ul
  className="accordion"
>
  <li
    className="accordion-item-1"
    key="0"
  >
    <h2
      className="_header"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      <AccessibleImage
        className="arrow expanded"
      />
      <span
        className="header-label"
      >
        Test Accordion Item 1
      </span>
    </h2>
    <div
      className="_content"
    >
      <div />
    </div>
  </li>
  <li
    className="accordion-item-2"
    key="1"
  >
    <h2
      className="_header"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      <AccessibleImage
        className="arrow "
      />
      <span
        className="header-label"
      >
        Test Accordion Item 2
      </span>
      <div
        className="header-buttons"
        tabIndex="-1"
      >
        <button />
      </div>
    </h2>
  </li>
  <li
    className="accordion-item-3"
    key="2"
  >
    <h2
      className="_header"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      <AccessibleImage
        className="arrow expanded"
      />
      <span
        className="header-label"
      >
        Test Accordion Item 3
      </span>
    </h2>
    <div
      className="_content"
    >
      <div />
    </div>
  </li>
</ul>
`;
